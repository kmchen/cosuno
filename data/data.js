export const data =
{
   "company":[
      {
         "name":"AUTO1 Group",
         "Logo": "https://source.unsplash.com/random",
         "specialty": ["ecommerce", "software"],
         "city": "Hamburg"
      },
      {
         "name":"Blacklane",
         "Logo": "https://source.unsplash.com/random",
         "specialty": ["fintech", "software"],
         "city": "Frankfurt"
      },
      {
         "name":"Blinkist",
         "Logo": "https://source.unsplash.com/random",
         "specialty": ["fintech", "software"],
         "city": "Munich"
      },
      {
         "name":"Bonify",
         "Logo": "https://source.unsplash.com/random",
         "specialty": ["fintech", "software"],
         "city": "Berlin"
      },
      {
         "name":"Camunda",
         "Logo": "https://source.unsplash.com/random",
         "specialty": ["ecommerce", "software"],
         "city": "Cologne"
      },
      {
         "name":"ChartMogul",
         "Logo": "https://source.unsplash.com/random",
         "specialty": ["ecommerce", "software"],
         "city": "Munich"
      },
      {
         "name":"Contentful",
         "Logo": "https://source.unsplash.com/random",
         "specialty": ["Software"],
         "city": "Cologne"
      },
      {
         "name":"Cosuno",
         "Logo": "https://source.unsplash.com/random",
         "specialty": ["construction", "software"],
         "city": "Berlin"
      },
      {
         "name":"Coya",
         "Logo": "https://source.unsplash.com/random",
         "specialty": ["ecommerce", "software"],
         "city": "Hamburg"
      },
      {
         "name":"Crate.io",
         "Logo": "https://source.unsplash.com/random",
         "specialty": ["fintech", "software"],
         "city": "Munich"
      },
      {
         "name":"Dalia Research",
         "Logo": "https://source.unsplash.com/random",
         "specialty": ["ecommerce", "software"],
         "city": "Munich"
      },
      {
         "name":"door2door",
         "Logo": "https://source.unsplash.com/random",
         "specialty": ["fintech", "software"],
         "city": "Frankfurt"
      },
      {
         "name":"Eyeo",
         "Logo": "https://source.unsplash.com/random",
         "specialty": ["ecommerce", "software"],
         "city": "Cologne"
      },
      {
         "name":"FreightHub",
         "Logo": "https://source.unsplash.com/random",
         "specialty": ["fintech", "software"],
         "city": "Frankfurt"
      },
      {
         "name":"German Autolabs",
         "Logo": "https://source.unsplash.com/random",
         "specialty": ["fintech", "software"],
         "city": "Frankfurt"
      },
      {
         "name":"GetYourGuide",
         "Logo": "https://source.unsplash.com/random",
         "specialty": ["fintech", "software"],
         "city": "Dusseldorf"
      },
      {
         "name":"GoEuro",
         "Logo": "https://source.unsplash.com/random",
         "specialty": ["fintech", "software"],
         "city": "Frankfurt"
      },
      {
         "name":"Helping",
         "Logo": "https://source.unsplash.com/random",
         "specialty": ["ecommerce", "software"],
         "city": "Hamburg"
      },
      {
         "name":"Infarm",
         "Logo": "https://source.unsplash.com/random",
         "specialty": ["fintech", "software"],
         "city": "Munich"
      },
      {
         "name":"Junique",
         "Logo": "https://source.unsplash.com/random",
         "specialty": ["ecommerce", "software"],
         "city": "Hamburg"
      },
      {
         "name":"MoBerries",
         "Logo": "https://source.unsplash.com/random",
         "specialty": ["fintech", "software"],
         "city": "Munich"
      },
      {
         "name":"Morressier",
         "Logo": "https://source.unsplash.com/random",
         "specialty": ["fintech", "software"],
         "city": "Munich"
      },
      {
         "name":"Movinga",
         "Logo": "https://source.unsplash.com/random",
         "specialty": ["fintech", "software"],
         "city": "Frankfurt"
      },
      {
         "name":"N26",
         "Logo": "https://source.unsplash.com/random",
         "specialty": ["fintech", "software"],
         "city": "Berlin"
      },
      {
         "name":"PEAT",
         "Logo": "https://source.unsplash.com/random",
         "specialty": ["fintech", "software"],
         "city": "Munich"
      },
      {
         "name":"PPRO",
         "Logo": "https://source.unsplash.com/random",
         "specialty": ["fintech", "software"],
         "city": "Berlin"
      },
      {
         "name":"Quentic",
         "Logo": "https://source.unsplash.com/random",
         "specialty": ["ecommerce", "software"],
         "city": "Cologne"
      },
      {
         "name":"ResearchGate",
         "Logo": "https://source.unsplash.com/random",
         "specialty": ["fintech", "software"],
         "city": "Munich"
      },
      {
         "name":"Skoove",
         "Logo": "https://source.unsplash.com/random",
         "specialty": ["fintech", "software"],
         "city": "Munich"
      },
      {
         "name":"SnapShot",
         "Logo": "https://source.unsplash.com/random",
         "specialty": ["ecommerce", "software"],
         "city": "Munich"
      },
      {
         "name":"Watchmaster",
         "Logo": "https://source.unsplash.com/random",
         "specialty": ["ecommerce", "software"],
         "city": "Hamburg"
      },
      {
         "name":"Wundertax",
         "Logo": "https://source.unsplash.com/random",
         "specialty": ["fintech", "software"],
         "city": "Berlin"
      },
      {
         "name":"Zeotap",
         "Logo": "https://source.unsplash.com/random",
         "specialty": ["ecommerce", "software"],
         "city": "Munich"
      },
   ]
};
