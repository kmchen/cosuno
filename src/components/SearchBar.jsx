import React from 'react';
import { Subject } from 'rxjs';
import { map } from 'rxjs/operators';
import PropTypes from 'prop-types';
import Autosuggest from 'react-autosuggest';
import match from 'autosuggest-highlight/match';
import parse from 'autosuggest-highlight/parse';
import TextField from '@material-ui/core/TextField';
import Paper from '@material-ui/core/Paper';
import MenuItem from '@material-ui/core/MenuItem';
import { makeStyles } from '@material-ui/core/styles';
import Container from '@material-ui/core/Container';

import Filters from './Filters';

const onSearch$ = new Subject();

const useStyles = makeStyles(theme => ({
  root: {
    flexGrow: 1,
  },
  container: {
    position: 'relative',
  },
  heroContent: {
    backgroundColor: '#f5f5f5',
    padding: theme.spacing(8, 0, 6),
  },
  suggestionsContainerOpen: {
    position: 'absolute',
    zIndex: 1,
    marginTop: theme.spacing(1),
    left: 0,
    right: 0,
  },
  suggestion: {
    display: 'block',
  },
  suggestionsList: {
    margin: 0,
    padding: 0,
    listStyleType: 'none',
  },
  divider: {
    height: theme.spacing(2),
  },
}));

function renderInputComponent(inputProps) {
  const { classes, inputRef = () => {}, ref, ...other } = inputProps;
  return (
    <TextField
      fullWidth
      InputProps={{
        inputRef: node => {
          ref(node);
          inputRef(node);
        },
        classes: {
          input: classes.input,
        },
      }}
      {...other}
    />
  );
}

function renderSuggestion(suggestion, { query, isHighlighted }) {
  const matches = match(suggestion.name, query);
  const parts = parse(suggestion.name, matches);
  return (
    <MenuItem selected={isHighlighted} component="div">
      <div>
        {parts.map(part => (
          <span key={part.text} style={{ fontWeight: part.highlight ? 500 : 400 }}>
            {part.text}
          </span>
        ))}
      </div>
    </MenuItem>
  );
}

function getSuggestionValue(suggestion) {
  return suggestion.name;
}

function onSuggestionSelected(dispatch) {
  /* eslint-disable no-unused-vars */
  return (event, { suggestion, suggestionValue }) => {
    dispatch(suggestionValue);
  };
}

export default function SearchBar({ fuse, setResult }) {
  const classes = useStyles();
  const [state, setState] = React.useState({
    company: '',
    filters: {
      fintech: false,
      construction: false,
      ecommerce: false,
      software: false,
    },
  });

  const [stateSuggestions, setSuggestions] = React.useState([]);

  const subscription = onSearch$
    .pipe(
      map(val => fuse.search(val)),
      map(companies => {
        return companies.reduce((acc, curr, idx) => {
          if (curr.specialty.filter(spec => state.filters[spec.toLowerCase()]).length) {
            acc.push(curr);
          }
          if (idx === companies.length - 1 && acc.length === 0) {
            const hasFilter = Object.keys(state.filters).reduce(
              (accFilters, currFilter) => accFilters || state.filters[currFilter],
              false,
            );
            return hasFilter ? acc : companies;
          }
          return acc;
        }, []);
      }),
    )
    .subscribe(results => {
      setSuggestions(results);
    });

  React.useEffect(() => {
    return () => {
      subscription.unsubscribe();
    };
  }, []);

  const handleSuggestionsFetchRequested = ({ value }) => {
    const inputValue = value.trim().toLowerCase();
    onSearch$.next(inputValue);
  };

  const handleSuggestionsClearRequested = () => {
    setSuggestions([]);
  };

  const handleChange = name => (event, { newValue }) => {
    setState({
      ...state,
      [name]: newValue,
    });
  };

  const handleFilter = name => event => {
    setState({
      ...state,
      filters: {
        ...state.filters,
        [name]: event.target.checked,
      },
    });
  };

  const autosuggestProps = {
    renderInputComponent,
    suggestions: stateSuggestions,
    onSuggestionsFetchRequested: handleSuggestionsFetchRequested,
    onSuggestionsClearRequested: handleSuggestionsClearRequested,
    getSuggestionValue,
    renderSuggestion,
    onSuggestionSelected: onSuggestionSelected(setResult),
  };

  return (
    <div className={classes.heroContent}>
      <Container maxWidth="sm">
        <div className={classes.root}>
          <Filters stateFilters={state.filters} handleFilter={handleFilter} />
          <Autosuggest
            {...autosuggestProps}
            inputProps={{
              classes,
              id: 'react-autosuggest-simple',
              label: 'Search for a company?',
              placeholder: 'try Cosuno?',
              value: state.company,
              onChange: handleChange('company'),
            }}
            theme={{
              container: classes.container,
              suggestionsContainerOpen: classes.suggestionsContainerOpen,
              suggestionsList: classes.suggestionsList,
              suggestion: classes.suggestion,
            }}
            renderSuggestionsContainer={options => (
              <Paper {...options.containerProps} square>
                {options.children}
              </Paper>
            )}
          />
        </div>
      </Container>
    </div>
  );
}

/* eslint-disable react/forbid-prop-types */
SearchBar.propTypes = {
  fuse: PropTypes.object.isRequired,
  setResult: PropTypes.func.isRequired,
};
