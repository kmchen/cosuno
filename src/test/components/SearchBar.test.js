import React from 'react';
import { shallow, configure } from 'enzyme';
import Fuse from 'fuse.js';
import Adapter from 'enzyme-adapter-react-16';
import SearchBar from '../../components/SearchBar';
import { data } from '../../../data/data.js';

configure({ adapter: new Adapter() });

const options = {
  shouldSort: true,
  threshold: 0.0,
  location: 0,
  distance: 100,
  maxPatternLength: 32,
  minMatchCharLength: 1,
  keys: ['name'],
};

describe('SearchBar component', () => {
  let SearchBarComponent;

  beforeEach(() => {
    const props = {
      fuse: new Fuse(data, options),
      setResult: jest.fn(),
    };
    SearchBarComponent = shallow(<SearchBar {...props} />);
  });

  it('should render SearchBar component', () => {
    expect(SearchBarComponent).toMatchSnapshot();
  });

  it('should render AutuSuggest component', () => {
    const autoSuggest = SearchBarComponent.find('Autosuggest');
    expect(autoSuggest.exists()).toBeTruthy();
  });
});
