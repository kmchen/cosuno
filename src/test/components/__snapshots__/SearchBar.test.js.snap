// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchBar component should render SearchBar component 1`] = `
<div
  className="makeStyles-heroContent-3"
>
  <WithStyles(ForwardRef(Container))
    maxWidth="sm"
  >
    <div
      className="makeStyles-root-1"
    >
      <Filters
        handleFilter={[Function]}
        stateFilters={
          Object {
            "construction": false,
            "ecommerce": false,
            "fintech": false,
            "software": false,
          }
        }
      />
      <Autosuggest
        alwaysRenderSuggestions={false}
        focusInputOnSuggestionClick={true}
        getSuggestionValue={[Function]}
        highlightFirstSuggestion={false}
        id="1"
        inputProps={
          Object {
            "classes": Object {
              "container": "makeStyles-container-2",
              "divider": "makeStyles-divider-7",
              "heroContent": "makeStyles-heroContent-3",
              "root": "makeStyles-root-1",
              "suggestion": "makeStyles-suggestion-5",
              "suggestionsContainerOpen": "makeStyles-suggestionsContainerOpen-4",
              "suggestionsList": "makeStyles-suggestionsList-6",
            },
            "id": "react-autosuggest-simple",
            "label": "Search for a company?",
            "onChange": [Function],
            "placeholder": "try Cosuno?",
            "value": "",
          }
        }
        multiSection={false}
        onSuggestionSelected={[Function]}
        onSuggestionsClearRequested={[Function]}
        onSuggestionsFetchRequested={[Function]}
        renderInputComponent={[Function]}
        renderSuggestion={[Function]}
        renderSuggestionsContainer={[Function]}
        shouldRenderSuggestions={[Function]}
        suggestions={Array []}
        theme={
          Object {
            "container": "makeStyles-container-2",
            "suggestion": "makeStyles-suggestion-5",
            "suggestionsContainerOpen": "makeStyles-suggestionsContainerOpen-4",
            "suggestionsList": "makeStyles-suggestionsList-6",
          }
        }
      />
    </div>
  </WithStyles(ForwardRef(Container))>
</div>
`;
